knitr/convert-odes/convert_odes.Rmd

```diff
@@ -0,0 +1,398 @@
+`ode_rk45`. Because the states are handled as `vector` variables, the solver
+output is an array of vectors (`vector[]`).
+
+```{stan, output.var = "", eval = FALSE}
+transformed parameters {
+  vector<lower=0>[4] y[N_t];
+  {
+    y = ode_rk45(simple_SIR, y0, t0, t, beta, kappa, gamma, xi, delta);
+  }
+}
+```
+
+### Full Model
+
+The full model with the new interface can be found
+[here](https://github.com/stan-dev/stat_comp_benchmarks/blob/master/benchmarks/sir/sir.stan);
+the full model with the old interface can be found
+[here](https://github.com/stan-dev/stat_comp_benchmarks/blob/master/benchmarks/sir/sir.stan),
+and the data can be found
+[here](https://github.com/stan-dev/stat_comp_benchmarks/blob/master/benchmarks/sir/sir.data.R).
+
+## PKPD Model
+
+### ODE System Function
+
+In the old system function, parameters are manually unpacked from `theta` and
+`x_r`:
+
+```
+functions {
+  real[] one_comp_mm_elim_abs(real t,
+                              real[] y,
+                              real[] theta,
+                              real[] x_r,
+                              int[] x_i) {
+    real dydt[1];
+    real k_a = theta[1]; // Dosing rate in 1/day
+    real K_m = theta[2]; // Michaelis-Menten constant in mg/L
+    real V_m = theta[3]; // Maximum elimination rate in 1/day
+    real D = x_r[1];
+    real V = x_r[2];
+    real dose = 0;
+    real elim = (V_m / V) * y[1] / (K_m + y[1]);
+
+    if (t > 0)
+      dose = exp(- k_a * t) * D * k_a / V;
+
+    dydt[1] = dose - elim;
+
+    return dydt;
+  }
+}
+```
+
+In the new interface, they are passed directly and so the unpacking is avoided:
+
+```
+functions {
+  vector one_comp_mm_elim_abs(real t,
+                              vector y,
+                              real k_a, // Dosing rate in 1/day
+                              real K_m, // Michaelis-Menten constant in mg/L
+                              real V_m, // Maximum elimination rate in 1/day
+                              real D,
+                              real V) {
+    vector[1] dydt;
+
+    real dose = 0;
+    real elim = (V_m / V) * y[1] / (K_m + y[1]);
+
+    if (t > 0)
+      dose = exp(- k_a * t) * D * k_a / V;
+
+    dydt[1] = dose - elim;
+
+    return dydt;
+  }
+}
+```
+
+### Calling the ODE Solver
+
+In the old interface the `theta` and `x_r` arguments are packed manually, and
+the `x_i` argument is required even though it isn't used:
+
+```
+transformed data {
+  real x_r[2] = {D, V};
+  int x_i[0];
+}
+...
+transformed parameters {
+  real C[N_t, 1];
+  {
+    real theta[3] = {k_a, K_m, V_m};
+    C = integrate_ode_bdf(one_comp_mm_elim_abs, C0, t0, times, theta, x_r, x_i);
+  }
+}
+```
+
+In the new interface the arguments are simply passed at the end of the
+`ode_bdf` call:
+
+```
+transformed parameters {
+  vector[1] C[N_t];
+  {
+    real theta[3] = {k_a, K_m, V_m};
+    C = ode_bdf(one_comp_mm_elim_abs, C0, t0, times, k_a, K_m, V_m, D, V);
```

Comment:
let's get rid of the inner block and do

`vector[1] C[N_t] = ode_bdf(one_comp_mm_elim_abs, C0, t0, times, k_a, K_m, V_m, D, V);`